modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderFactory.java

```
@@ -0,0 +1,253 @@
+  private Map<String, ServiceRegistration> providerRegistrations = new ConcurrentHashMap<String, ServiceRegistration>();;
+
+  /** The OSGI bundle context */
+  protected BundleContext bundleContext = null;
+
+  /** The organization directory service */
+  private OrganizationDirectoryService orgDirectory;
+
+  /** OSGi callback for setting the organization directory service. */
+  public void setOrgDirectory(OrganizationDirectoryService orgDirectory) {
+    this.orgDirectory = orgDirectory;
+  }
+
+  /**
+   * Callback for the activation of this component
+   *
+   * @param cc
+   *          the component context
+   */
+  public void activate(ComponentContext cc) {
+    logger.debug("Activate CanvasUserProviderFactory");
+    this.bundleContext = cc.getBundleContext();
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.osgi.service.cm.ManagedServiceFactory#getName()
+   */
+  @Override
+  public String getName() {
+    return PID;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.osgi.service.cm.ManagedServiceFactory#updated(java.lang.String, java.util.Dictionary)
+   */
+  @Override
+  public void updated(String pid, Dictionary properties) throws ConfigurationException {
+
+    logger.debug("updated CanvasUserProviderFactory");
+
+    String organization = (String) properties.get(ORGANIZATION_KEY);
+    if (StringUtils.isBlank(organization)) throw new ConfigurationException(ORGANIZATION_KEY, "is not set");
+
+    String url = (String) properties.get(CANVAS_URL_KEY);
+    if (StringUtils.isBlank(url)) throw new ConfigurationException(CANVAS_URL_KEY, "is not set");
+
+    String token = (String) properties.get(CANVAS_TOKEN_KEY);
+    if (StringUtils.isBlank(token)) throw new ConfigurationException(CANVAS_TOKEN_KEY, "is not set");
+
+    String courseIdentifierProperty = (String) properties.get(COURSE_IDENTIFIER_PROPERTY_KEY);
+    if (!"id".equals(courseIdentifierProperty) && !"sis_course_id".equals(courseIdentifierProperty)) {
```

if only `id` and `sis_course_id` is allowed here, I would strongly prefer to get an error if it is set to another value and I would only use the default if it is unset. The reasoning is that someone could set this to `sis_cuorse_id` (←note the typo) and this code would just silently use `id` instead. If you just fail fast instead, people will notice the problem and can solve it directly.